packages/headless/src/controllers/generated-answer/headless-generated-answer.ts

```diff
@@ -0,0 +1,139 @@
+import {GeneratedAnswerCitation} from '../../api/generated-answer/generated-answer-event-payload';
+import {SearchEngine} from '../../app/search-engine/search-engine';
+import {
+  streamAnswer,
+  resetAnswer,
+  likeGeneratedAnswer,
+  dislikeGeneratedAnswer,
+} from '../../features/generated-answer/generated-answer-actions';
+import {
+  logDislikeGeneratedAnswer,
+  logLikeGeneratedAnswer,
+  logOpenGeneratedAnswerSource,
+  logRetryGeneratedAnswer,
+} from '../../features/generated-answer/generated-answer-analytics-actions';
+import {generatedAnswerReducer as generatedAnswer} from '../../features/generated-answer/generated-answer-slice';
+import {GeneratedAnswerState} from '../../features/generated-answer/generated-answer-state';
+import {executeSearch} from '../../features/search/search-actions';
+import {GeneratedAnswerSection} from '../../state/state-sections';
+import {loadReducerError} from '../../utils/errors';
+import {buildController} from '../controller/headless-controller';
+
+export type {GeneratedAnswerState, GeneratedAnswerCitation};
+
+/**
+ * @internal
+ */
+export interface GeneratedAnswer {
+  /**
+   * The state of the GeneratedAnswer controller.
+   */
+  state: GeneratedAnswerState;
+  /**
+   * Re-executes the last query to generate an answer.
+   */
+  retry(): void;
+  /**
+   * Determines if the generated answer was liked, or upvoted by the end user.
+   */
+  like(): void;
+  /**
+   * Determines if the generated answer was disliked, or downvoted by the end user.
+   */
+  dislike(): void;
+  /**
+   * Logs a custom event indicating a cited source link was clicked.
+   * @param id The ID of the clicked citation.
+   */
+  logCitationClick(id: string): void;
+}
+
+/**
+ * @internal
+ */
+export function buildGeneratedAnswer(engine: SearchEngine): GeneratedAnswer {
+  if (!loadGeneratedAnswerReducer(engine)) {
+    throw loadReducerError;
+  }
+
+  const {dispatch} = engine;
+  const controller = buildController(engine);
+  const getState = () => engine.state;
+
+  let abortController: AbortController | undefined;
+  let lastRequestId: string;
+  let lastStreamId: string;
+
+  const setAbortControllerRef = (ref: AbortController) => {
+    abortController = ref;
+  };
+
+  const getIsStreamInProgress = () => {
+    if (!abortController || abortController?.signal.aborted) {
+      abortController = undefined;
+      return false;
+    }
+    return true;
+  };
+
+  const subscribeToSearchRequests = () => {
+    const strictListener = () => {
+      const state = getState();
+      const requestId = state.search.requestId;
+      const streamId =
+        state.search.extendedResults.generativeQuestionAnsweringId;
+
+      if (lastRequestId !== requestId) {
```

Comment:
When would this happen? In case of retry? A comment might be helpful.

Reply:
AFAIK this listener will run on every state update of which there are many that could not involve any change in the `requestId`. My condition here is just to execute this code when a new request has been made and processed by redux.